### ISSUE TITLE
`fallbackValue` is not applied for vararg multi-value options (was: Support fallbackValue with arity="0..*")

### ISSUE THREAD

**Issue by sewe** [remkop/picocli#1904]
Please support `fallbackValue` with "zero-based" arities other than `0..1`, in particular with `0..*`.

To see why this might be useful, consider the following `@Option` definition:

```java
enum DebugFacility { FOO, BAR, BAZ, ALL }

@Option(
    names = { "--debug" },
    paramLabel = "FACILITY", //
    split = ",",
    arity = "0..*", //
    fallbackValue = "ALL", //
    description = "The facilities for which to log debug information (one or more of ${COMPLETION-CANDIDATES}). "
        + "If none are given, debug information will be logged for ${FALLBACK-VALUE} categories.")
public void setDebugFacilities(Collection<DebugFacility> facilities) { ... }
```

This should support the following scenarios

```
--debug FOO               -> FOO
--debug FOO --debug BAR   -> FOO, BAR
--debug FOO,BAR           -> FOO, BAR
--debug all               -> ALL (i.e., FOO, BAR, BAZ)
--debug                   -> ALL (i.e., FOO, BAR, BAZ)
```

The last one unfortunately doesn't work, as the `fallbackValue` does not get passed to `setDebugFacilities`; it merely gets an empty `facilities` `Collection`.

And we can't insert `ALL` programmatically in `setDebugFacilities`, either. To see why, consider the following scenario:

```
--debug FOO --debug       -> ALL (i.e., FOO, BAR, BAZ)
```

Here, the parameterless `--debug` simply results in another call of `setDebugFacilities` with `FOO` as single parameter. In other words, one cannot detect the parameterless case

Maybe there is a workaround that I am unable to see, but IMHO `fallbackValue` support for any "zero-based" arity might be the most idiomatic solution. WDYT?

**Comment by remkop** [comment]
I believe all the examples mentioned would work when the option is defined with `arity = "0..1"`. 
That would allow you to make your use case work with the current version of picocli (unless I’m missing something).

I still need to think more about the general use case you mentioned…

**Comment by remkop** [comment]
What comes to mind for the general case is that in the setter method, the logic could compare the parameter collection with the current value (the field where the collection is stored) and if both collections are equal, the end user didn’t specify a parameter. The logic can then programmatically add the fallback value. 

I haven’t tested this though, does it work? 😅

**Comment by remkop** [comment]
One more thing: I generally recommend against using `arity = "0..*"` because it introduces parser ambiguity when the application also has positional parameters. 

The question to ask is “does my application really need to support cases like the below?”

```
--debug FOO BAR BAZ
```

For the cases you mentioned above, arity 0..1 is sufficient. It still has some potential ambiguity. If this ambiguity becomes a problem, please check the section on [Custom Parameter Processing](https://picocli.info/#_custom_parameter_processing).

**Comment by sewe** [comment]
> The question to ask is “does my application really need to support cases like the below?”
> ```
> --debug FOO BAR BAZ
> ```

No, that is not a case I need to support, but maybe I misunderstood `arity`, then.

Interestingly, I can use an `arity` of `0..1` on a `Collection` valued setter.

```java
@Option(
    names = { "--debug" },
    paramLabel = "FACILITY", //
    split = ",",
    arity = "0..1", //
    fallbackValue = "ALL", //
    description = "The facilities for which to log debug information (one or more of ${COMPLETION-CANDIDATES}). "
        + "If none are given, debug information will be logged for ${FALLBACK-VALUE} categories.")
public void setDebugFacilities(Collection<DebugFacility> facilities) { ... }
```

And then picocli still accepts the following as before (unit tests for the win!).

```
--debug FOO --debug BAR 
--debug FOO,BAR
```

Apparently, my mental model of `arity` is wrong. 😞 

But even when changing `arity` to `0..1`, picocli doesn't seem apply the `fallbackValue`, at least in some cases.

```
--debug /some/positional/parameter/and/most/certainly/not/an/enum/value
```

calls `setDebugFacilities` with an empty list but

```
--debug -- /some/positional/parameter/and/most/certainly/not/an/enum/value
```

works and injects the `fallbackValue` of `ALL`. It does call `setDebugFacilities` twice, though. 🤯 

I just don't get why picocli wouldn't, in the first case, either throw an `Invalid value for option '--debug: /some/positional/parameter/and/most/certainly/not/an/enum/value` or, preferrably, pass the `fallthroughValue`. An empty `facilities` `Collection` seems to be wrong:

> If an option is defined with arity = "0..1", it [may or may not have a parameter value](https://picocli.info/#_optional_values). If such an option is specified without a value on the command line, it is assigned the fallback value.

**Comment by sewe** [comment]
> What comes to mind for the general case is that in the setter method, the logic could compare the parameter collection with the current value (the field where the collection is stored) and if both collections are equal, the end user didn’t specify a parameter. The logic can then programmatically add the fallback value.

> I haven’t tested this though, does it work? 😅 

Not without further ugliness.

ATM I treat `setDebugFacilities` as a setter, i.e., it completely replaces the existing `this.facilities` as, AFAICT, picocli keeps its own list of `FACILITY` values so future calls pass not only the values of the current option but of all previously parsed options, too.

Unfortunately, this means I cannot simply do this:

```java
public void setDebugFacilities(Set<DebugFacility> facilities) {
    if (facilities.equals(this.debugCategories)) {
        this.facilities = DebugFacility.values();
    } else {
        this.facilities = facilities;
    }
}
```

Even in the simplest case

```
--debug /some/positional/parameter/and/most/certainly/not/an/enum/value
```

`setDebugFacilities` is called twice (why?) and in both cases without parameters. Now, as intended the first call updates `this.facilities` to contain all values. Alas, as picocli keeps its own list of `FACILITY` values, the _second_ call is still parameterless and then I end up in the `else` case.

Maybe a workaround is still possible by refining this further and being even more stateful in the `@Option` handler, but somehow that feels wrong. After all, the nice thing about picocli is its declarativeness.

**Comment by remkop** [comment]
> 
> Apparently, my mental model of `arity` is wrong. 😞

Picocli's concept of arity is documented in the user manual: https://picocli.info/#_arity 
It is the number of parameters _for each option occurrence_.
For example, arity=2 would mean that the end user must specify 2 parameters after the option name: `--option 1 2`.

> 
> But even when changing `arity` to `0..1`, picocli doesn't seem apply the `fallbackValue`, at least in some cases.
> 
> ```
> --debug /some/positional/parameter/and/most/certainly/not/an/enum/value
> ```
> 
> calls `setDebugFacilities` with an empty list but
> 
> ```
> --debug -- /some/positional/parameter/and/most/certainly/not/an/enum/value
> ```
> 
> works and injects the `fallbackValue` of `ALL`. It does call `setDebugFacilities` twice, though. 🤯

Just a note that what we are encountering here is the parser ambiguity I mentioned earlier. A [parameter preprocessor](https://picocli.info/#IParameterPreprocessor_Parser_Plugin) may be needed to resolve this.

> 
> I just don't get why picocli wouldn't, in the first case, either throw an `Invalid value for option '--debug: /some/positional/parameter/and/most/certainly/not/an/enum/value` or, preferrably, pass the `fallthroughValue`. An empty `facilities` `Collection` seems to be wrong:
> 
> > If an option is defined with arity = "0..1", it [may or may not have a parameter value](https://picocli.info/#_optional_values). If such an option is specified without a value on the command line, it is assigned the fallback value.

Not sure about the 2 calls or about why the setter is called with an empty collection when the value cannot be converted to a `DebugFacility` instance. Is `DebugFacility` an enum or do you have a custom type converter? (And if the latter, what does this converter do for invalid values? Does it throw an exception?)

You can see a bit more what is going on internally by switching on tracing: https://picocli.info/#_tracing
and looking at the output for this case:
```
--debug /some/positional/parameter/and/most/certainly/not/an/enum/value
```

Can you post the output here?

**Comment by sewe** [comment]
Sorry for the late replay; I was away for the holidays.

> You can see a bit more what is going on internally by switching on tracing: https://picocli.info/#_tracing
and looking at the output for this case:
> ```
> --debug /some/positional/parameter/and/most/certainly/not/an/enum/value
> ```
> Can you post the output here?

I created a test that reproduces this issue: https://github.com/sewe/picocli/commit/7525c9cc06f2d97ac3b18e84c6f0f84844ce5ca8 (feel free to copy it)

Running the only test case which fails (`singleOptionWithFallback`) with `DEBUG` tracing yields the following:

```
[picocli DEBUG] Creating CommandSpec for picocli.Issue1904$TestCommand@490caf5f with factory picocli.CommandLine$DefaultFactory
[picocli INFO] Picocli version: 4.7.1-SNAPSHOT, JVM: 17.0.4 (Debian OpenJDK 64-Bit Server VM 17.0.4+8-Debian-1deb11u1), OS: Linux 5.10.0-19-amd64 amd64
[picocli INFO] Parsing 2 command line args [--debug, example.txt]
[picocli DEBUG] Parser configuration: optionsCaseInsensitive=false, subcommandsCaseInsensitive=false, abbreviatedOptionsAllowed=false, abbreviatedSubcommandsAllowed=false, allowOptionsAsOptionParameters=false, allowSubcommandsAsOptionParameters=false, aritySatisfiedByAttachedOptionParam=false, atFileCommentChar=#, caseInsensitiveEnumValuesAllowed=false, collectErrors=false, endOfOptionsDelimiter=--, expandAtFiles=true, limitSplit=false, overwrittenOptionsAllowed=false, posixClusteredShortOptionsAllowed=true, separator=null, splitQuotedStrings=false, stopAtPositional=false, stopAtUnmatched=false, toggleBooleanFlags=false, trimQuotes=false, unmatchedArgumentsAllowed=false, unmatchedOptionsAllowedAsOptionParameters=true, unmatchedOptionsArePositionalParams=false, useSimplifiedAtFiles=false
[picocli DEBUG] (ANSI is disabled by default: systemproperty[picocli.ansi]=null, isatty=false, TERM=xterm-256color, OSTYPE=null, isWindows=false, JansiConsoleInstalled=false, ANSICON=null, ConEmuANSI=null, NO_COLOR=null, CLICOLOR=null, CLICOLOR_FORCE=null)
[picocli DEBUG] Initializing command 'null' (user object: picocli.Issue1904$TestCommand@490caf5f): 1 options, 1 positional parameters, 1 required, 0 groups, 0 subcommands.
[picocli DEBUG] Set initial value for field java.util.Set<picocli.Issue1904$DebugFacility> picocli.Issue1904$TestCommand.debugFacilities of type interface java.util.Set to [].
[picocli DEBUG] Set initial value for field java.util.List<java.io.File> picocli.Issue1904$TestCommand.paths of type interface java.util.List to null.
[picocli DEBUG] [0] Processing argument '--debug'. Remainder=[example.txt]
[picocli DEBUG] '--debug' cannot be separated into <option>=<option-parameter>
[picocli DEBUG] Found option named '--debug': field java.util.Set<picocli.Issue1904$DebugFacility> picocli.Issue1904$TestCommand.debugFacilities, arity=0..1
[picocli DEBUG] Split with regex ',' resulted in 1 parts: [example.txt]
[picocli DEBUG] example.txt cannot be assigned to option --debug: type conversion fails: Invalid value for option '--debug': expected one of [FOO, BAR, BAZ, ALL] (case-sensitive) but was 'example.txt'.
[picocli DEBUG] Initializing binding for option '--debug' (FACILITY) on TestCommand@490caf5f with empty Set
[picocli DEBUG] [1] Processing argument 'example.txt'. Remainder=[]
[picocli DEBUG] 'example.txt' cannot be separated into <option>=<option-parameter>
[picocli DEBUG] Could not find option 'example.txt', deciding whether to treat as unmatched option or positional parameter...
[picocli DEBUG] No option named 'example.txt' found. Processing as positional parameter
[picocli DEBUG] 'example.txt' doesn't resemble an option: 0 matching prefix chars out of 1 option names
[picocli DEBUG] [1] Processing next arg as a positional parameter. Command-local position=0. Remainder=[example.txt]
[picocli DEBUG] Position 0 (command-local) is in index range 0..*. Trying to assign args to field java.util.List<java.io.File> picocli.Issue1904$TestCommand.paths, arity=1..*
[picocli DEBUG] 'example.txt' doesn't resemble an option: 0 matching prefix chars out of 1 option names
[picocli INFO] Adding [example.txt] to field java.util.List<java.io.File> picocli.Issue1904$TestCommand.paths for args[0..*] at position 0 on TestCommand@490caf5f
[picocli DEBUG] Initializing binding for positional parameter at index 0..* (<paths>) on TestCommand@490caf5f with empty List
[picocli DEBUG] Consumed 1 arguments and 0 interactive values, moving command-local position to index 1.
[picocli DEBUG] Applying default values for command '<main class>'
```

Hope this helps.

At any rate, the test also nicely illustrates all cases that already _do_ work, which is most of them 😃.

**Comment by remkop** [comment]
Thanks for posting this.

After looking at the debug trace and the code, I see now that this is an existing bug.
This bug does not manifest for single-value options, but does manifest for collections or maps (multi-value) options.

### Analysis
The parser tries to determine for "vararg" options whether the next arg is an option parameter or not.
If the option has a fallbackValue, and the parser can determine in advance that the next arg is not an option parameter, the fallbackValue is inserted before the next arg. So far, so good, but...

The problem is that picocli has a quick check (`varargCanConsumeNextValue`) and a thorough check (`canConsumeOneArgument` and `canConsumeOneMapArgument`) to determine whether an arg is an option parameter.
The thorough check includes parameter value splitting and type conversion.

The fallbackValue is only applied if the quick check fails.   **<-- This is the bug...**

In this use case, the quick check passes, but the thorough check (the type conversion to enum value) fails.
Because of the bug, the fallbackValue is not applied in this case.

### Solution
The fix is also perform the thorough check when determining whether to apply the fallbackValue.

**Comment by remkop** [comment]
The fix has been pushed to the main branch and will be included in the upcoming 4.7.1 release.
Thank you again for raising this!
